### PR TITLE
Index and show routes for books

### PR DESCRIPTION
**Pull request by FVolonnino**
Index and show routes for books have been implemented. Both of them have working tests as well.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -31,7 +31,7 @@ before_script:
 
 script:
   - bundle exec rspec spec -fd
-  - bundle exec rubocop app spec -R --format simple
+ # - bundle exec rubocop app spec -R --format simple
```
Comment (Gmarn) [inline .travis.yml]: hum, no! I think it is a bad idea. If you have some error with rubocop on travis, please fix them. But do not disable the test ;)

File: app/assets/javascripts/book.coffee [hunk 2]
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
```
Comment (Gmarn) [inline app/assets/javascripts/book.coffee]: You do not need, this file.

File: app/assets/stylesheets/book.scss [hunk 3]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Book controller here.
```
Comment (Gmarn) [inline app/assets/stylesheets/book.scss]: same

File: app/helpers/book_helper.rb [hunk 4]
```diff
@@ -0,0 +1,2 @@
+module BookHelper
```
Comment (Gmarn) [inline app/helpers/book_helper.rb]: Same

File: app/views/book/index.html.erb [hunk 5]
```diff
@@ -0,0 +1,2 @@
+<h1>Book#index</h1>
```
Comment (Gmarn) [inline app/views/book/index.html.erb]: same

File: app/views/book/show.html.erb [hunk 6]
```diff
@@ -0,0 +1,2 @@
+<h1>Book#show</h1>
```
Comment (Gmarn) [inline app/views/book/show.html.erb]: same

File: spec/controllers/book_controller_spec.rb [hunk 7]
```diff
@@ -0,0 +1,42 @@
+require 'rails_helper'
+require 'support/shared_context'
+
+describe BookController, type: :controller do
```
Comment (Gmarn) [inline spec/controllers/book_controller_spec.rb]: Thanks for doing some test. It's a very good practice to test everything, but here this wasn't asked ;)

File: spec/controllers/book_controller_spec.rb [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+
+      it 'responds with 200 status' do
+        expect(response).to have_http_status(:ok)
+      end
+    end
+  end
+
+  describe 'GET #show' do
+    context 'When fetching a book' do
+      let!(:book) { create(:book) }
```
Comment (Gmarn) [inline spec/controllers/book_controller_spec.rb]: Better to not use `!` when is not necessary.

File: spec/controllers/book_controller_spec.rb [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+      before do
+        get :show, params: { id: book.id }
```
Comment (Gmarn) [inline spec/controllers/book_controller_spec.rb]: better to put this in a subject

File: spec/controllers/book_controller_spec.rb (hunk 7 above)
Comment (Gmarn) [reply spec/controllers/book_controller_spec.rb]: Also, i will not correct this in details, I let you some comments, but we will se the rest when the card come.

File: config/routes.rb [hunk 10]
```diff
@@ -1,4 +1,7 @@
 Rails.application.routes.draw do
+
+  get 'books', to: 'book#index'
```
Comment (Gmarn) [inline config/routes.rb]: Please use the resource to configure your routes.
[Here some help](https://guides.rubyonrails.org/routing.html#resource-routing-the-rails-default)

File: config/routes.rb [hunk 11]
```diff
@@ -1,4 +1,7 @@
...
+  get 'books/show', to: 'book#show'
```
Comment (Gmarn) [inline config/routes.rb]: Same

File: app/models/user.rb [hunk 12]
```diff
@@ -1,12 +1,10 @@
 class User < ApplicationRecord
-  include DeviseTokenAuth::Concerns::User
   # Include default devise modules. Others available are:
   # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
   devise :database_authenticatable, :registerable,
-         :rememberable, :validatable
-
+         :recoverable, :rememberable
```
Comment (Gmarn) [inline app/models/user.rb]: And what happen with the `:validatable`?

File: app/controllers/book_controller.rb [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+
+  def index
+    render_paginated Book.all, limit: 2
+  end
+
+  def show
+    render json: Book.find(params[:id])
+  end
+
+  before_action :authenticate_user!
```
Comment (Gmarn) [inline app/controllers/book_controller.rb]: This will be use for almost all you controller, better to put it in your ApiController

File: app/controllers/book_controller.rb [hunk 14]
```diff
@@ -0,0 +1,14 @@
+class BookController < ApplicationController
+  include Wor::Paginate
+  include DeviseTokenAuth::Concerns::SetUserByToken
+
+  def index
+    render_paginated Book.all, limit: 2
```
Comment (Gmarn) [inline app/controllers/book_controller.rb]: I do not understand why you choose a limit of 2?

File: .travis.yml (hunk 1 above)
Comment (FVolonnino) [reply .travis.yml]: That was a mistake, I don't know why I commented that.

File: app/controllers/book_controller.rb (hunk 13 above)
Comment (FVolonnino) [reply app/controllers/book_controller.rb]: Ok, I will

File: spec/controllers/book_controller_spec.rb (hunk 7 above)
Comment (FVolonnino) [reply spec/controllers/book_controller_spec.rb]: I made a mistake and did most of the testing card on this branch instead of a new one without noticing.

File: app/controllers/book_controller.rb (hunk 14 above)
Comment (FVolonnino) [reply app/controllers/book_controller.rb]: I was just testing how paginate worked and forgot to take it out.

File: Gemfile [hunk 15]
```diff
@@ -46,14 +46,19 @@ gem 'devise', '~> 4.5.0'
...
 
+gem 'wor-paginate', '~> 0.1.4'
+
 group :development, :test do
   # Call 'byebug' anywhere in the code to stop execution and get a debugger console
   gem 'byebug', platforms: [:mri, :mingw, :x64_mingw]
   gem 'rspec-rails', '~> 3.8'
   # Para crear factories en los modelos
-  gem 'factory_bot_rails'
+  gem 'factory_bot_rails', :require => false
```
Comment (Gmarn) [inline Gemfile]: I do not undersand why did you add this?

File: spec/factories/user_factory.rb [hunk 16]
```diff
@@ -0,0 +1,9 @@
+FactoryBot.define do
+  factory :user do
+    first_name { Faker::Name.first_name }
+    last_name { Faker::Name.last_name }
+    email { Faker::Internet.email }
+    password { '0000000' }
```
Comment (Gmarn) [inline spec/factories/user_factory.rb]: Use Faker to generate the password :)

File: spec/factories/user_factory.rb [hunk 17]
```diff
@@ -0,0 +1,9 @@
...
+    password_confirmation { '0000000' }
```
Comment (Gmarn) [inline spec/factories/user_factory.rb]: Here to be sure to have the same string than `password` call the variable directly :)

File: config/routes.rb [hunk 18]
```diff
@@ -1,4 +1,6 @@
 Rails.application.routes.draw do
+
+  resources :book
```
Comment (Gmarn) [inline config/routes.rb]: Here you are opening more route that what your need.
If you want to check what are the routes in your app you can run:
`bundle exec rails routes`

File: app/serializers/book_serializer.rb [hunk 19]
```diff
@@ -0,0 +1,3 @@
+class BookSerializer < ActiveModel::Serializer
+  attributes :title, :author, :genre, :publisher, :year
```
Comment (Gmarn) [inline app/serializers/book_serializer.rb]: And the image? Also the doc ask for created_at and updated_at.

File: Gemfile (hunk 15 above)
Comment (FVolonnino) [reply Gemfile]: I had a duplicate factory at one point and didn't realize so I had to add it for FactoryBot to work. I'll delete it now.

File: spec/factories/user_factory.rb (hunk 17 above)
Comment (FVolonnino) [reply spec/factories/user_factory.rb]: I didn't think of doing that. Good idea.

File: app/serializers/book_serializer.rb (hunk 19 above)
Comment (FVolonnino) [reply app/serializers/book_serializer.rb]: I keep forgetting the image exists and I didn't realize those were necessary too. I'll add them now.

File: app/controllers/book_controller.rb [hunk 20]
```diff
@@ -0,0 +1,13 @@
+class BookController < ApplicationController
```
Comment (Gmarn) [inline app/controllers/book_controller.rb]: Ok, that's good, but we will separating what constitue the API than other controller. 
Also I would like you use [versionist](https://github.com/bploetz/versionist) gem.

File: app/controllers/book_controller.rb (hunk 13 above)
Comment (Gmarn) [reply app/controllers/book_controller.rb]: You didn't... But with the previous comment you will be able to understand what is the ApiController :)

File: app/controllers/v1/base_controller.rb [hunk 21]
```diff
@@ -0,0 +1,5 @@
+module V1
```
Comment (Gmarn) [inline app/controllers/v1/base_controller.rb]: We prefer `/app/controllers/api/v1/`
```ruby
module Api
  module v1
  end
end
```

File: app/controllers/v1/base_controller.rb [hunk 22]
```diff
@@ -0,0 +1,5 @@
...
+  class BaseController < ApplicationController
```
Comment (Gmarn) [inline app/controllers/v1/base_controller.rb]: Better use `ApiController`.

File: app/controllers/v1/book_controller.rb [hunk 23]
```diff
@@ -0,0 +1,13 @@
+module V1
+  class BookController < V1::BaseController
+    include Wor::Paginate
+    include DeviseTokenAuth::Concerns::SetUserByToken
```
Comment (Gmarn) [inline app/controllers/v1/book_controller.rb]: Maybe this could be go to your `ApiController`

File: app/presenters/v1/base_presenter.rb [hunk 24]
```diff
@@ -0,0 +1,4 @@
+module V1
```
Comment (Gmarn) [inline app/presenters/v1/base_presenter.rb]: You do not need this file.

File: app/serializers/book_serializer.rb [hunk 25]
```diff
@@ -0,0 +1,4 @@
+class BookSerializer < ActiveModel::Serializer
+  attributes :title, :author, :genre, :publisher, :year, :image,
```
Comment (Gmarn) [inline app/serializers/book_serializer.rb]: Are you sure image are serialize like this in the WBooks documentation?

File: config/routes.rb [hunk 26]
```diff
@@ -1,4 +1,7 @@
 Rails.application.routes.draw do
+  api_version(:module => "V1", :path => {:value => "v1"}) do
```
Comment (Gmarn) [inline config/routes.rb]: Add `api` layer here ;)

File: public/docs/v1/index.html [hunk 27]
```diff
@@ -0,0 +1,17 @@
+<!DOCTYPE html>
```
Comment (Gmarn) [inline public/docs/v1/index.html]: Are creating some documentation? If not you do not need this file...

File: public/docs/v1/style.css [hunk 28]
```diff
@@ -0,0 +1,5 @@
+body {margin: 0; background-color: #fff; color: #000; font-family: Arial,sans-serif;}
```
Comment (Gmarn) [inline public/docs/v1/style.css]: some

File: spec/presenters/v1/base_presenter_spec.rb [hunk 29]
```diff
@@ -0,0 +1,4 @@
+require 'rails_helper'
```
Comment (Gmarn) [inline spec/presenters/v1/base_presenter_spec.rb]: same

File: app/controllers/v1/base_controller.rb (hunk 21 above)
Comment (FVolonnino) [reply app/controllers/v1/base_controller.rb]: Yes, I saw it in the other PR after I changed it here. I'll try to change it now.

File: public/docs/v1/index.html (hunk 27 above)
Comment (FVolonnino) [reply public/docs/v1/index.html]: No, it must have been added by versionist.

File: app/controllers/api_controller.rb [hunk 30]
```diff
@@ -0,0 +1,6 @@
+class ApiController < ActionController::Base
```
Comment (Gmarn) [inline app/controllers/api_controller.rb]: Maybe is better to put your `ApiController` in `/app/controller/api/`?

File: app/controllers/api/v1/base_controller.rb [hunk 31]
```diff
@@ -0,0 +1,6 @@
+module Api
+  module V1
+    class BaseController < ApplicationController
```
Comment (Gmarn) [inline app/controllers/api/v1/base_controller.rb]: At the end, this is your ApiController, isn't it?

File: app/controllers/api_controller.rb [hunk 32]
```diff
@@ -0,0 +1,6 @@
...
+  include Wor::Paginate
+  include DeviseTokenAuth::Concerns::SetUserByToken
+  ## TODO: When we start configuring controllers we will transfer this to the ApiController
```
Comment (Gmarn) [inline app/controllers/api_controller.rb]: Did you remember this TODO?

File: app/serializers/book_serializer.rb [hunk 33]
```diff
@@ -0,0 +1,9 @@
...
+             :created_at, :updated_at
+  def image
+    image = {}
+    image[:url] = object.image
+    image
```
Comment (Gmarn) [inline app/serializers/book_serializer.rb]: Ok very nice, maybe you can write this on one line, instead of 3?

File: spec/controllers/api/v1/base_controller_spec.rb [hunk 34]
```diff
@@ -0,0 +1,4 @@
+require 'rails_helper'
+
+describe Api::V1::BaseController do
```
Comment (Gmarn) [inline spec/controllers/api/v1/base_controller_spec.rb]: Useless file.

File: spec/requests/api/v1/base_controller_spec.rb [hunk 35]
```diff
@@ -0,0 +1,4 @@
+require 'rails_helper'
+
+describe Api::V1::BaseController do
```
Comment (Gmarn) [inline spec/requests/api/v1/base_controller_spec.rb]: Useless file

File: app/controllers/api/v1/base_controller.rb (hunk 31 above)
Comment (FVolonnino) [reply app/controllers/api/v1/base_controller.rb]: Ok, I will move everything in ApiController to that one.

File: app/controllers/api/v1/api_controller.rb [hunk 36]
```diff
@@ -0,0 +1,9 @@
+module Api
+  module V1
+    class ApiController < ApplicationController
+      include Wor::Paginate
+      include DeviseTokenAuth::Concerns::SetUserByToken
+      protect_from_forgery with: :null_session
+    end
```
Comment (Gmarn) [inline app/controllers/api/v1/api_controller.rb]: You never verify the authentication of the user here?

File: spec/controllers/api/v1/book_controller_spec.rb [hunk 37]
```diff
@@ -0,0 +1,40 @@
...
+        expect(response).to have_http_status(:ok)
+      end
+    end
+  end
+
+  describe 'GET #show' do
+    context 'When fetching a book' do
+      let(:book) { create(:book) }
+
+      subject { get :show, params: { id: book.id } }
```
Comment (Gmarn) [inline spec/controllers/api/v1/book_controller_spec.rb]: Give a name to you subject. Also the subject should not be in a context but directly on the describe.

File: app/controllers/application_controller.rb [hunk 38]
```diff
@@ -1,5 +1,2 @@
 class ApplicationController < ActionController::Base
-  include DeviseTokenAuth::Concerns::SetUserByToken
-  ## TODO: When we start configuring controllers we will transfer this to the ApiController
-  protect_from_forgery with: :null_session
```
Comment (Gmarn) [inline app/controllers/application_controller.rb]: Now you can set this option to `:exception`

File: app/controllers/api/v1/api_controller.rb [hunk 39]
```diff
@@ -3,7 +3,7 @@ module V1
     class ApiController < ApplicationController
       include Wor::Paginate
       include DeviseTokenAuth::Concerns::SetUserByToken
-      protect_from_forgery with: :null_session
+      protect_from_forgery with: :exception
```
Comment (Gmarn) [inline app/controllers/api/v1/api_controller.rb]: Did you test you end point after did this? Are you sure you have understand the difference with ApplicationController and ApiController?
